optuna/study/_tell.py

```diff
@@ -22,59 +20,96 @@
 _logger = logging.get_logger(__name__)
 
 
-def _check_and_convert_to_values(
-    n_objectives: int, original_value: Optional[Union[float, Sequence[float]]], trial_number: int
-) -> Tuple[Optional[List[float]], Optional[str]]:
-    if isinstance(original_value, Sequence):
-        _original_values: Sequence[Optional[float]] = list(original_value)
+def _get_frozen_trial(study: "optuna.Study", trial: Union[trial_module.Trial, int]) -> FrozenTrial:
+    if isinstance(trial, trial_module.Trial):
+        trial_id = trial._trial_id
+    elif isinstance(trial, int):
+        trial_number = trial
+        try:
+            trial_id = study._storage.get_trial_id_from_study_id_trial_number(
+                study._study_id, trial_number
+            )
+        except NotImplementedError as e:
+            warnings.warn(
+                "Study.tell may be slow because the trial was represented by its number but "
+                f"the storage {study._storage.__class__.__name__} does not implement the "
+                "method required to map numbers back. Please provide the trial object "
+                "to avoid performance degradation."
+            )
+
+            trials = study.get_trials(deepcopy=False)
+
+            if len(trials) <= trial_number:
+                raise ValueError(
+                    f"Cannot tell for trial with number {trial_number} since it has not been "
+                    "created."
+                ) from e
+
+            trial_id = trials[trial_number]._trial_id
+        except KeyError as e:
+            raise ValueError(
+                f"Cannot tell for trial with number {trial_number} since it has not been "
+                "created."
+            ) from e
+    else:
+        raise TypeError("Trial must be a trial object or trial number.")
+
+    return study._storage.get_trial(trial_id)
+
+
+def _check_state_and_values(
+    state: Optional[TrialState], values: Optional[Union[float, Sequence[float]]]
+) -> None:
+    if state == TrialState.COMPLETE:
+        if values is None:
+            raise ValueError(
+                "No values were told. Values are required when state is TrialState.COMPLETE."
+            )
+    elif state in (TrialState.PRUNED, TrialState.FAIL):
+        if values is not None:
+            raise ValueError(
+                "Values were told. Values cannot be specified when state is "
+                "TrialState.PRUNED or TrialState.FAIL."
+            )
+    elif state is not None:
+        raise ValueError(f"Cannot tell with state {state}.")
+
+
+def _check_values(
```

Comment:
[Nit] I feel the roles of `_check_state_and_values` and `_check_values` are unclear at a glance. How about renaming this function to something like `_check_non_none_values`? Or do you come up with any better idea?

Reply:
How about `_check_values_are_feasible`?

Reply:
That sounds fine! (Or you can also ignore this comment because now I feel it's too minor a point...)